go/vt/vttablet/tabletmanager/restore.go

```diff
@@ -166,6 +188,166 @@ func (tm *TabletManager) restoreDataLocked(ctx context.Context, logger logutil.L
 	return nil
 }
 
+// restoreToTimeFromBinlog restores to the snapshot time of the keyspace
+// currently this works with mysql based database only (as it uses mysql specific queries for restoring
+func (tm *TabletManager) restoreToTimeFromBinlog(ctx context.Context, pos mysql.Position, restoreTime *vttime.Time) error {
+	// validate the dependent settings
+	if *binlogHost == "" || *binlogPort <= 0 || *binlogUser == "" {
+		log.Warning("invalid binlog server setting, restoring to last available backup.")
+		return nil
+	}
+
+	timeoutCtx, cancelFnc := context.WithTimeout(ctx, *timeoutForGTIDLookup)
+	defer cancelFnc()
+	afterGTIDPos, beforeGTIDPos := tm.getGTIDFromTimestamp(timeoutCtx, pos, restoreTime.Seconds)
+	if afterGTIDPos == "" {
+		return vterrors.New(vtrpcpb.Code_FAILED_PRECONDITION, fmt.Sprintf("unable to fetch the GTID for the specified time - %s", restoreTime.String()))
+	}
+
+	log.Infof("going to restore upto the GTID - %s", afterGTIDPos)
+	if beforeGTIDPos == "" {
+		beforeGTIDPos = pos.GTIDSet.Last()
+	}
+	err := tm.catchupToGTID(timeoutCtx, afterGTIDPos, beforeGTIDPos)
+	if err != nil {
+		return vterrors.Wrapf(err, "unable to replicate upto desired GTID : %s", afterGTIDPos)
+	}
+
+	return nil
+}
+
+// getGTIDFromTimestamp gets the next GTID of the event happened on the timestamp (resore_to_time)
+//
+// it returns the 2 values, 1st one is the after gtid of the timestamp,
+// the 2nd one returns before gtid upto which the replication will be applied
+// 1st can be used directly in the query `START SLAVE UNTIL SQL_BEFORE_GTIDS = ''`
+// 2nd will be used to check if replication completed from the binlog server
+func (tm *TabletManager) getGTIDFromTimestamp(ctx context.Context, pos mysql.Position, restoreTime int64) (string, string) {
+	connParams := &mysql.ConnParams{
+		Host:  *binlogHost,
+		Port:  *binlogPort,
+		Uname: *binlogUser,
+	}
+	if binlogPwd != nil && *binlogPwd != "" {
+		connParams.Pass = *binlogPwd
+	}
+	dbCfgs := &dbconfigs.DBConfigs{
+		Host: connParams.Host,
+		Port: connParams.Port,
+	}
+	dbCfgs.SetDbParams(*connParams, *connParams)
+	vsClient := vreplication.NewReplicaConnector(connParams)
+
+	filter := &binlogdatapb.Filter{
+		Rules: []*binlogdatapb.Rule{{
+			Match: "/.*",
+		}},
+	}
+
+	gtidsChan := make(chan []string)
+	beforeGTIDPos := ""
+	currentPos := ""
+
+	go func() {
+		err := vsClient.VStream(ctx, mysql.EncodePosition(pos), filter, func(events []*binlogdatapb.VEvent) error {
+			for _, event := range events {
+				if event.Gtid != "" {
+					currentPos = event.Gtid
+					if event.Timestamp > restoreTime {
```

Comment:
```suggestion
					if event.Timestamp >= restoreTime {
```

Reply:
I think this is important. In the design we say that we will catch up to the last event before restoreTime.